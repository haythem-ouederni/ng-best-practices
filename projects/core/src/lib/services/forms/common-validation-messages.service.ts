import {Injectable} from '@angular/core';
import {ValidationErrors} from '@angular/forms';
/**
 * CommonValidatorMessagesService must be provided at each using module.
 * And tha calling module must load the common/fr.json file in addiction
 * to its own traduction files
 *
 * IMPORTANT : at the actual state of the service no need to add specific
 * unit test for it. Orignally, it simply returns a constant with fixed value.
 * It was transformed to a service just to be able to use the translation.
 *
 */
@Injectable({
  providedIn: 'root',
})
export class CommonValidationMessagesService {
  private _validationMessages: {[key: string]: any};

  get validationMessages() {
    if (!this._validationMessages) {
      this.initializeValidationMessage();
    }
    return this._validationMessages;
  }

  /**
   * Method called to initialize the common validation messages
   */
  private initializeValidationMessage() {
    this._validationMessages = {
      required: {
        required: 'Required input',
      },
      invalid: {
        pattern: 'Invalid input',
      },
      length: {
        minlength: {
          minlength: (validationErrors: ValidationErrors) => `The minimal length is ${validationErrors.requiredLength}`,
        },
        maxlength: {
          maxlength: (validationErrors: ValidationErrors) => `The maximal length is ${validationErrors.requiredLength}`,
        },
      },
      numbers: {
        decimal: {
          pattern: {
            pattern: 'Should be a number',
          },
        },
        integer: {
          pattern: {
            pattern: 'Should be an integer',
          },
        },
        min: {
          min: (validationErrors: any) => `The minimal value is ${validationErrors.min}`,
        },
        max: {
          max: (validationErrors: any) => `The maximal value is ${validationErrors.max}`,
        },
      },
    };
  }
}
