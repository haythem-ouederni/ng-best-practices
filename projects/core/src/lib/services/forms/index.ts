export * from './common-validation-messages.service';
