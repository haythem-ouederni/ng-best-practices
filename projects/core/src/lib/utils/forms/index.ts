export * from './form-errors.util';
