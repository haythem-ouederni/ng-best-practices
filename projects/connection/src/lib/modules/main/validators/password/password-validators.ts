import {AbstractControl, ValidationErrors, ValidatorFn} from '@angular/forms';

export class PasswordValidators {
  static beginWithValidator(prefix: string): ValidatorFn {
    return (control: AbstractControl): {[key: string]: any} | null => {
      const beginsWith = control.value && control.value.indexOf(prefix) === 0;
      return !beginsWith ? {beginsWith: {value: control.value, prefix}} : null;
    };
  }
}

export class PasswordValidatorsMessages {
  static beginsWith = {
    beginsWith: (validationErrors: ValidationErrors) =>
      validationErrors.value
        ? `${validationErrors.value} must begin with ${validationErrors.prefix}`
        : `Must begin with ${validationErrors.prefix}`,
  };
}
