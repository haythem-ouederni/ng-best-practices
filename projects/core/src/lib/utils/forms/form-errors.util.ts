import {FormGroup} from '@angular/forms';
import {FormGroupFinalErrorsMessages, FormGroupValidationErrorsMessages, ErrorMessageFunction} from '../../models';

export class FormErrorsUtil {
  /**
   * Aggregates multiple validation errors messages.
   * For each formControl it will return a list of the related errors message
   * @param form : formGroup
   * @param validationErrorsMessages : An object containing the different errors messages for each field of the formgroup
   * @return a formErrors
   *
   * example of a validationMessages:
   * validationMessages = {
   *      age: {
   *          required: 'Age required.',
   *          min: (validationErrors: ValidationErrors) => `Age at least ${validationErrors.min}`,
   * }
   *
   */
  static aggregateErrorsMessages(
    form: FormGroup,
    validationErrorsMessages: FormGroupValidationErrorsMessages
  ): FormGroupFinalErrorsMessages {
    const formErrors: FormGroupFinalErrorsMessages = {};
    if (form && validationErrorsMessages) {
      Object.keys(validationErrorsMessages)
        .filter((key: string) => form.controls[key] && !form.controls[key].pending && form.controls[key].invalid)
        .forEach((key: string) => {
          formErrors[key] = [];
          Object.keys(form.controls[key].errors).forEach((error: string) => {
            let msg = '';
            if (typeof validationErrorsMessages[key][error] === 'function') {
              const msgFunction: ErrorMessageFunction = validationErrorsMessages[key][error] as ErrorMessageFunction;
              msg = msgFunction(form.controls[key].errors[error]);
            } else {
              msg = validationErrorsMessages[key][error] as string;
            }
            if (formErrors[key] && formErrors[key].length > 0) {
              formErrors[key].push(msg);
            } else {
              formErrors[key] = [msg];
            }
          });
        });
    }
    return formErrors;
  }
}
