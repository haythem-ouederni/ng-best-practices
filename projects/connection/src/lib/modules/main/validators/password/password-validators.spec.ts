import {async} from '@angular/core/testing';
import {FormControl, ValidationErrors} from '@angular/forms';
import {PasswordValidators, PasswordValidatorsMessages} from './password-validators';

const prefix = 'my-prefix';

describe('PasswordValidators', () => {
  describe('#beginWithValidator', () => {
    test('should return null for valid control', async(() => {
      const formValue = `${prefix}-my-value`;
      const validFormControl: FormControl = new FormControl(formValue, PasswordValidators.beginWithValidator(prefix));
      expect(validFormControl.errors).toEqual(null);
    }));

    test('should return the invalid control information', () => {
      const formValue = `uncorrect-value`;
      const expectedErrors = {
        beginsWith: {
          value: formValue,
          prefix,
        },
      };
      const validFormControl: FormControl = new FormControl(formValue, PasswordValidators.beginWithValidator(prefix));
      expect(validFormControl.errors).toEqual(expectedErrors);
    });
  });
});

describe('PasswordValidatorsMessages', () => {
  describe('#beginsWith', () => {
    test('should match snapshot when value inquired', () => {
      const value = 'my-value';
      const validationErrors: ValidationErrors = {
        value,
        prefix,
      };
      expect(PasswordValidatorsMessages.beginsWith.beginsWith(validationErrors)).toMatchSnapshot();
    });
    test('should match snapshot when value not inquired', () => {
      const validationErrors: ValidationErrors = {
        prefix,
      };
      expect(PasswordValidatorsMessages.beginsWith.beginsWith(validationErrors)).toMatchSnapshot();
    });
  });
});
