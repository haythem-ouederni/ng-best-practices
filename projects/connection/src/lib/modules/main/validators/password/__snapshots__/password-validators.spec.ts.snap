// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordValidatorsMessages #beginsWith should match snapshot when value inquired 1`] = `"my-value must begin with my-prefix"`;

exports[`PasswordValidatorsMessages #beginsWith should match snapshot when value not inquired 1`] = `"Must begin with my-prefix"`;
